packages/framework/ai-collab/src/explicit-strategy/agentEditReducer.ts

```diff
@@ -0,0 +1,499 @@
+		}
+	}
+}
+
+function isNodeAllowedType(node: TreeNode, allowedTypes: TreeNodeSchema[]): boolean {
+	for (const allowedType of allowedTypes) {
+		if (Tree.is(node, allowedType)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+function isPrimitive(content: unknown): boolean {
+	return (
+		typeof content === "number" ||
+		typeof content === "string" ||
+		typeof content === "boolean" ||
+		content === undefined ||
+		content === null
+	);
+}
+
+function isObjectTarget(selection: Selection): selection is ObjectTarget {
+	return Object.keys(selection).length === 1 && "target" in selection;
+}
+
+function isRange(selection: Selection): selection is Range {
+	return "from" in selection && "to" in selection;
+}
+
+interface RangeInfo {
+	array: TreeArrayNode;
+	startIndex: number;
+	endIndex: number;
+}
+
+function getRangeInfo(range: Range, idGenerator: IdGenerator): RangeInfo {
+	const { array: arrayFrom, index: startIndex } = getPlaceInfo(range.from, idGenerator);
+	const { array: arrayTo, index: endIndex } = getPlaceInfo(range.to, idGenerator);
+
+	if (arrayFrom !== arrayTo) {
+		throw new UsageError(
+			'The "from" node and "to" nodes of the range must be in the same parent array.',
+		);
+	}
+
+	return { array: arrayFrom, startIndex, endIndex };
+}
+
+function getPlaceInfo(
+	place: ObjectPlace | ArrayPlace,
+	idGenerator: IdGenerator,
+): {
+	array: TreeArrayNode;
+	index: number;
+} {
+	if (place.type === "arrayPlace") {
+		const parent = idGenerator.getNode(place.parentId) ?? fail("Expected parent node");
+		const child = (parent as unknown as Record<string, unknown>)[place.field];
+		if (child === undefined) {
+			throw new UsageError(`No child under field field`);
+		}
+		const schema = Tree.schema(child as TreeNode);
+		if (schema.kind !== NodeKind.Array) {
+			throw new UsageError("Expected child to be in an array node");
+		}
+		return {
+			array: child as TreeArrayNode,
+			index: place.location === "start" ? 0 : (child as TreeArrayNode).length,
+		};
+	} else {
+		const node = getNodeFromTarget(place, idGenerator);
+		const nodeIndex = Tree.key(node);
+		const parent = Tree.parent(node);
+		if (parent === undefined) {
+			throw new UsageError("TODO: root node target not supported");
+		}
+		const schema = Tree.schema(parent);
+		if (schema.kind !== NodeKind.Array) {
+			throw new UsageError("Expected child to be in an array node");
+		}
+		return {
+			array: parent as unknown as TreeArrayNode,
+			index: place.place === "before" ? (nodeIndex as number) : (nodeIndex as number) + 1,
+		};
+	}
+}
+
+/**
+ * Returns the target node with the matching internal objectId using the provided {@link ObjectTarget}
+ */
+function getNodeFromTarget(target: ObjectTarget, idGenerator: IdGenerator): TreeNode {
+	const node = idGenerator.getNode(target.target);
+	assert(node !== undefined, "objectId does not exist in nodeMap");
+	return node;
+}
+
+function objectIdsExist(treeEdit: TreeEdit, idGenerator: IdGenerator): void {
```

Comment:
Looks like this is a validation function, looking issues and reporting errors. I would recommend naming this something like `assertObjectIdsExist` (we generally recommend verb phrases for functions - see https://github.com/microsoft/FluidFramework/wiki/Coding-Guidelines#-do-name-functions-using-verb-phrases). It would probably be a good idea to document this as well.

Reply:
Added JSDOC and updated function naming in commit 36ed33dd3f0a4f456cbf9f5ea2024f6573bb91b9